pyiso/nspower.py

```diff
@@ -0,0 +1,275 @@
+import pandas
+import pytz
+from datetime import timedelta, datetime
+from pyiso import LOGGER
+from pyiso.base import BaseClient
+
+
+class NSPowerClient(BaseClient):
+    """
+    Nova Scotia Power (Canada) has a public, informational page titled "Today's Power" which is backed by several
+    useful JSON data sources.
+
+    https://www.nspower.ca/en/home/about-us/todayspower.aspx
+
+    This client provides:
+        * historical loads up to 24 hours in the past and load forecasts up to 24 hours into the future.
+        * historical generation mix up to 24 hours in the past through to the start of the current hour.
+    """
+    NAME = 'NSP'
+    TZ_NAME = 'Canada/Atlantic'
+
+    # LM6000s are combined cycle gas turbines. I don't know if the value being listed seperately represents just the
+    # condensing steam generator (i.e. Tuft Cove 6) or the entire combined cycle system of two natural gas generators
+    # (Tuft Cove 4 & 5) plus the condensing steam generator (Tuft Cove 6).
+    # See http://www.novascotia.ca/nse/ea/tuftscove6/NSPI-TuftsCove6-Registration.pdf
+    #
+    # CTs are diesel (oil) combustion turbines (I think).
+    # See http://www.nspower.ca/en/home/about-us/how-we-make-electricity/thermal-electricity/oil-facilities.aspx
+    #
+    # "HFO/Natural Gas" is Heavy Fuel Oil/Natural Gas
+    fuels = {
+        'Solid Fuel': 'coal',
+        'HFO/Natural Gas': 'natgas',
+        'CT\'s': 'oil',
+        'LM 6000\'s': 'thermal',
```

Comment:
@xinyue-luna, what is the meaning behind the `thermal` fuel type from `BaseClient`? It it geothermal or steam turbines? Nova Scotia Power's [LM6000 units](https://en.wikipedia.org/wiki/General_Electric_LM6000) are the second stage (condensing steam) of a combined cycle generation with two natural gas generators. See this [Nova Scotia Power proposal document](http://www.novascotia.ca/nse/ea/tuftscove6/NSPI-TuftsCove6-Registration.pdf) for context. How should the `BaseClient` fuel choices reflect combined cycle generation?

There are more fuel choices returned by the JSON than are reflected in https://www.nspower.ca/en/home/about-us/todayspower.aspx I would be interested to know how they aggregate all the returned fuels in that page's javascript.

Reply:
Thermal refers to undefined fossil fuel type, especially the case where there is a mix of natural gas and coal, but the ratio is unclear. Fossil, as you might have noticed, is used in NYISO, and it actually represents coal. 
Geothermal is 'geo' in pyiso. 
CHP is considered as natgas in pyiso for now.
Regarding fuel aggregation by Nova Scotia, would you mind sending me the json file? Thanks!

Reply:
> Regarding fuel aggregation by Nova Scotia, would you mind sending me the json file? 

The [`currentmix.json`](https://github.com/r24mille/pyiso/blob/bc2e61e527bc06180ac2e78368e9fc01c918366c/tests/fixtures/nsp/currentmix.json) file I added to the unit test fixtures gives a fully example. I'll copy the element for one timestamp here for discussion:

```
"datetime":  "\/Date(1507114800000)\/",
"Solid Fuel":  41.99,
"HFO/Natural Gas":  13.08,
"CT\u0027s":  0.07,
"LM 6000\u0027s":  9.59,
"Biomass":  0.98,
"Hydro":  2.74,
"Wind":  30.64,
"Imports": 0.92
```

The section of code that this comment and my other comment in this pull request is the mapping of Nova Scotia Power's fule names to pyiso fuel names. Here it is in its entirety:

```
fuels = {
    'Solid Fuel': 'coal',
    'HFO/Natural Gas': 'natgas',
    'CT\'s': 'oil',
    'LM 6000\'s': 'thermal',
    'Biomass': 'biomass',
    'Hydro': 'hydro',
    'Wind': 'wind',
    'Imports': 'other'
}
```

My justifications are:
* Solid Fuel : `coal`- The term "Solid Fuel" is used to describe both coal and petcoke. Though the percentages described in their [Thermal Electricity](https://www.nspower.ca/en/home/about-us/how-we-make-electricity/thermal-electricity/default.aspx) page suggest this value is primarily their coal facilities.
* HFO/Natural Gas : `natgas`- This stands for Heavy Fuel Oil/Natural Gas
* CT's : `oil`- I think that CTs are diesel (oil) combustion turbines. I based this interpretation on their [Oil Facilities](http://www.nspower.ca/en/home/about-us/how-we-make-electricity/thermal-electricity/oil-facilities.aspx) page which describes the three combustion turbines in their service region.
* LM 6000's : `thermal` - As discussed above, my interpretation is based on [the Tuft's Cove 6 power proposal document I found](http://www.novascotia.ca/nse/ea/tuftscove6/NSPI-TuftsCove6-Registration.pdf).
* Biomass : `biomass`
* Hydro : `hydro`
* Wind : `wind`
* Imports : `other` - This relates to my [other comment](https://github.com/WattTime/pyiso/pull/161#discussion_r143331357) on this pull request.

So, given your feedback, that CHP is currently maps to `natgas` in the pyiso fuel types. I propose one of two solutions:
1. We add `chp` as a base fuel type for pyiso and the `NSPowerClient` uses it.
2. In the parser, add both the "HFO/Natural Gas" value and "LM 6000's" value together in a single row for the `get_generation(...)` results since they both map to the same pyiso fuel value.

I prefer option 1 above, but I don't feel strongly about that.

On any unrelated note:

> Fossil, as you might have noticed, is used in NYISO, and it actually represents coal.

Should we open an issue to fix this in NYISO? Change mappings for NYISO's `fossil` value to `coal`?

Reply:
So, for what it's worth, Nova Scotia Power is lumping several of the fuels together under "Natural Gas" in their page https://www.nspower.ca/en/home/about-us/todayspower.aspx Curious which ones, I took a dive into the javascript backing that page: https://www.nspower.ca/site/renewables/assets/js/site.js?v=882528822

Lines 156-161:
```    
chartdata['repm'][repmCoal]['value'] = Math.round(mix['Solid Fuel']);
chartdata['repm'][repmNaturalGas]['value'] = Math.round(mix['CT\u0027s'] + mix['HFO/Natural Gas'] + mix['LM 6000\u0027s']);
chartdata['repm'][repmBiomass]['value'] = Math.round(mix['Biomass']);
chartdata['repm'][repmHydro]['value'] = Math.round(mix['Hydro']);
chartdata['repm'][repmWind]['value'] = Math.round(mix['Wind']);
chartdata['repm'][repmImports]['value'] = Math.round(mix['Imports']);
```

They're lumping CTs, HFO/Natural Gas, and LM 6000s together under the title "Natural Gas." Do we want to do the same or keep the granularity?